unix/vxworks: make DirEntry slightly smaller

cuviper opened the pull request:
`DirEntry` contains a `ReadDir` handle, which used to just be a wrapper
on `Arc<InnerReadDir>`. Commit af75314ecdbc5 added `end_of_stream: bool`
which is not needed by `DirEntry`, but adds 8 bytes after padding. We
can let `DirEntry` have an `Arc<InnerReadDir>` directly to avoid that.

rust-highfive commented:
r? @dtolnay

(rust_highfive has picked a reviewer for you, use r? to override)

dtolnay commented:
@bors r+

bors commented:
:pushpin: Commit 12c94b27f3c703eede3ab7621c5e8eaaa7ef4daa has been approved by `dtolnay`

<!-- @bors r=dtolnay 12c94b27f3c703eede3ab7621c5e8eaaa7ef4daa -->
<!-- homu: {"type":"Approved","sha":"12c94b27f3c703eede3ab7621c5e8eaaa7ef4daa","approver":"dtolnay"} -->

bors commented:
:hourglass: Testing commit 12c94b27f3c703eede3ab7621c5e8eaaa7ef4daa with merge 13edb9ca1cd1b2f50b7cb0d685d3db5ed59c35ec...
<!-- homu: {"type":"BuildStarted","head_sha":"12c94b27f3c703eede3ab7621c5e8eaaa7ef4daa","merge_sha":"13edb9ca1cd1b2f50b7cb0d685d3db5ed59c35ec"} -->

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/runs/1223994823)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/runs/1223994823","builder_name":"checks-actions"} -->

cuviper commented:
Fixed a clone that I had missed...

@bors r=dtolnay

bors commented:
:pushpin: Commit 5f762014b8f87f23742ec75ac22f9903b95276bc has been approved by `dtolnay`

<!-- @bors r=dtolnay 5f762014b8f87f23742ec75ac22f9903b95276bc -->
<!-- homu: {"type":"Approved","sha":"5f762014b8f87f23742ec75ac22f9903b95276bc","approver":"dtolnay"} -->

jonas-schievink commented:
@bors r- rollup=iffy failed in https://github.com/rust-lang/rust/pull/77712#issuecomment-705812486

cuviper commented:
Well that's annoying -- that `end_of_stream` has always been unused in practice on some targets, but the `derive(Clone)` that I removed was counting as a use. I've added more conditionals, but I'll pause for re-review just in case.

dtolnay commented on library/std/src/sys/unix/fs.rs:
```diff
@@ -944,7 +949,18 @@ pub fn readdir(p: &Path) -> io::Result<ReadDir> {
         } else {
             let inner = InnerReadDir { dirp: Dir(ptr), root };
-            Ok(ReadDir { inner: Arc::new(inner), end_of_stream: false })
+            cfg_if::cfg_if! {
+                if #[cfg(not(any(
+                    target_os = "solaris",
+                    target_os = "illumos",
+                    target_os = "fuchsia",
+                    target_os = "redox",
+                )))] {
+                    Ok(ReadDir { inner: Arc::new(inner), end_of_stream: false })
+                } else {
+                    Ok(ReadDir { inner: Arc::new(inner) })
+                }
+            }
```
```suggestion
            Ok(ReadDir {
                inner: Arc::new(inner),
                #[cfg(not(any(
                    target_os = "solaris",
                    target_os = "illumos",
                    target_os = "fuchsia",
                    target_os = "redox",
                )))]
                end_of_stream: false,
            })
```

dtolnay commented:
@bors r+ rollup=iffy

bors commented:
:pushpin: Commit 1d06b07765e1be71f3aeec1d0c0f365b0907c7a8 has been approved by `dtolnay`

<!-- @bors r=dtolnay 1d06b07765e1be71f3aeec1d0c0f365b0907c7a8 -->
<!-- homu: {"type":"Approved","sha":"1d06b07765e1be71f3aeec1d0c0f365b0907c7a8","approver":"dtolnay"} -->

bors commented:
:hourglass: Testing commit 1d06b07765e1be71f3aeec1d0c0f365b0907c7a8 with merge 5ddef544fa26c5535376fa424baf30cb985f7cb1...
<!-- homu: {"type":"BuildStarted","head_sha":"1d06b07765e1be71f3aeec1d0c0f365b0907c7a8","merge_sha":"5ddef544fa26c5535376fa424baf30cb985f7cb1"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/1233421681), [checks-azure](https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=37125)
Approved by: dtolnay
Pushing 5ddef544fa26c5535376fa424baf30cb985f7cb1 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"dtolnay","base_ref":"master","builders":{"checks-azure":"https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=37125","checks-actions":"https://github.com/rust-lang-ci/rust/runs/1233421681"},"merge_sha":"5ddef544fa26c5535376fa424baf30cb985f7cb1"} -->
